util/file_util.py

```diff
@@ -963,6 +997,121 @@ def file_copy_to_spreadsheet(filename: str,
+
+    Returns:
+      string with encoding such as 'utf8'
+    """
+    if rawdata is None:
+        rawdata = file_get_sample_bytes(file)
+    encoding_result = chardet.detect(rawdata)
+    if encoding_result:
+        encoding = encoding_result.get('encoding')
+        if encoding:
+            return encoding
+    return default
+
+
+def file_get_csv_reader_options(
+        file: str,
+        default_options: dict = {},
+        data: str = None,
+        encoding: str = None,
+        delim_chars: list = [',', '	', ';', '|', ':']) -> dict:
+    """Returns a dictionary with options for the CSV file reader.
+
+    Args:
+      file: name of the csv file to get encoding
+      default_options: default options returned if not detected
+        such as 'delimiter'.
+      data: string for which delimiter is to be detected
+        If data is not given, sample data is read from the file.
+      encoding: character encoding in the file.
+      delim_chars: list of possible delimiter characters.
+        If not set, non-alphanumeric characters from the first line
+        are used as candidate delimiter characters.
+
+    Returns:
+      dict with the following:
+        'delimiter': delimiter character for CSV files.
+        'dialect': File dialect, such as 'unix', 'excel'
+    """
+    result = dict(default_options)
+
+    if data is None:
+        # Get data from file decoded with the right encoding
+        rawdata = file_get_sample_bytes(file)
+        if encoding is None:
+            encoding = file_get_encoding(file, rawdata=rawdata)
+        data = rawdata.decode(encoding)
+
+    # Get the dialect for the data
+    try:
+        dialect = csv.Sniffer().sniff(data)
+    except csv.Error:
+        # Use default as excel as it may not be detected well.
+        dialect = 'excel'
+    if dialect:
+        result['dialect'] = dialect
+
+    # Get CSV delimiter by counting possible delimiter characters
+    # across rows and picking the most common delimiter.
+    rows = data.split('\n')
+    if not delim_chars:
+        # Get non alphanumeric characters from data.
+        delim_chars = {c for c in rows[0].strip() if not c.isalnum()}
+        logging.level_debug() and logging.debug(
+            f'Looking for delimiter in {file} among {delim_chars}')
+    char_counts = {c: [] for c in delim_chars}
+    for index in range(1, len(rows) - 1):
+        # Count possible delimiter characters per row
+        row = rows[index]
+        for char in delim_chars:
+            char_counts[char].append(row.count(char))
+    # Get the char with the same count across rows.
+    for c in char_counts.keys():
+        c_counts = char_counts[c]
+        if c_counts:
+            c_min = min(c_counts)
+            c_med = np.median(c_counts)
+            if c_min > 0 and c_min == c_med:
+                result['delimiter'] = c
+                break
+    logging.level_debug() and logging.debug(
```

Comment:
Logging level check could be skipped if we use logging.debug with arguments like 
logging.debug("Got options for file: %s, result = %s", file, result)

Reply:
Done. will try the AI tool to auto-correct these.